diff --git a/src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx b/src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx
index b3d58a5ff..e5a3889ba 100644
--- a/src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx
+++ b/src/drivers/Wayland/Fl_Wayland_Screen_Driver.cxx
@@ -229,6 +229,7 @@ static void pointer_enter(void *data, struct wl_pointer *wl_pointer, uint32_t se
   set_event_xy(win);
   need_leave = NULL;
   win = Fl_Wayland_Window_Driver::surface_to_window(surface);
+  seat->pointer_focus = surface;
   // Caution: with an Fl_Tooltip this call can hide the window being entered (#1317)
   if (!win->parent()) Fl::handle(FL_ENTER, win);
 }
@@ -834,7 +835,8 @@ static void wl_keyboard_key(void *data, struct wl_keyboard *wl_keyboard,
     Fl::handle(event, win);
   }
   if (event == FL_KEYDOWN && status == XKB_COMPOSE_NOTHING &&
-      !(sym >= FL_Shift_L && sym <= FL_Alt_R)) {
+      !(sym >= FL_Shift_L && sym <= FL_Alt_R) &&
+      !(sym >= FL_F && sym <= FL_F_Last)) {
     // Handling of key repeats :
     // Use serial argument rather than time to detect repeated keys because
     // serial value changes at each key up or down in all tested OS and compositors,
@@ -1574,6 +1576,7 @@ static bool compute_full_and_maximized_areas(Fl_Wayland_Screen_Driver::output *o
                                              int& Wworkarea, int& Hworkarea) {
   if (Fl_Wayland_Screen_Driver::compositor == Fl_Wayland_Screen_Driver::unspecified) {
     Wfullscreen = 0;
+    Fl::warning("Wayland compositor unknown");
     return false;
   }
   bool found_workarea = false;
@@ -1682,6 +1685,7 @@ int Fl_Wayland_Screen_Driver::x() {
   wl_list_for_each(output, &outputs, link) {
     break;
   }
+  assert(workarea_xywh[0] >= 0);
   return workarea_xywh[0] / (output->gui_scale * output->wld_scale);
 }
 
@@ -1692,6 +1696,7 @@ int Fl_Wayland_Screen_Driver::y() {
   wl_list_for_each(output, &outputs, link) {
     break;
   }
+  assert(workarea_xywh[1] >= 0);
   return workarea_xywh[1] / (output->gui_scale * output->wld_scale);
 }
 
@@ -1702,6 +1707,7 @@ int Fl_Wayland_Screen_Driver::w() {
   wl_list_for_each(output, &outputs, link) {
     break;
   }
+  assert(workarea_xywh[2] > 0);
   return workarea_xywh[2] / (output->gui_scale * output->wld_scale);
 }
 
@@ -1712,6 +1718,7 @@ int Fl_Wayland_Screen_Driver::h() {
   wl_list_for_each(output, &outputs, link) {
     break;
   }
+  assert(workarea_xywh[3] > 0);
   return workarea_xywh[3] / (output->gui_scale * output->wld_scale);
 }
 
